yjit/src/codegen.rs

```diff
@@ -4021,6 +4021,113 @@ fn jit_rb_false(
     true
 }
 
+
+/// Codegen for Kernel::is_a?
+fn jit_rb_kernel_is_a(
+    jit: &mut JITState,
+    ctx: &mut Context,
+    asm: &mut Assembler,
+    ocb: &mut OutlinedCb,
+    _ci: *const rb_callinfo,
+    _cme: *const rb_callable_method_entry_t,
+    _block: Option<IseqPtr>,
+    argc: i32,
+    known_recv_class: *const VALUE,
+) -> bool {
+    if argc != 1 {
+        return false;
+    }
+
+    // If this is a super class we might not know the class
+    if known_recv_class.is_null() {
+        return false;
+    }
+
+    // Important note: This works because we depend on known_recv_class.
+    // The left hand side here will always have the same type because there
+    // is a guard before this method call that ensures that is the case.
+    // We just guard that the right hand side does not change.
+
+    let sample_rhs = jit_peek_at_stack(jit, ctx, 0);
+    let sample_lhs = jit_peek_at_stack(jit, ctx, 1);
+
+    // If we are testing against a module, we can't be certain
+    // about the class hierarchy.
+    if !unsafe { RB_TYPE_P(sample_rhs, RUBY_T_CLASS) } {
```

Comment:
Don't you need to guard the same thing in generated code or have `jit_prepare_routine_call`? Since `rb_obj_is_kind_of` may raise for that case, just checking a compile-time instance seems not enough.

Reply:
It might be related to the CI failure too https://github.com/ruby/ruby/pull/7297/checks?check_run_id=11313649873

Reply:
We only call rb_obj_is_kind_of at comp time. And then we guard that `ctx.stack_opnd(0) == sample_rhs`. If the runtime value is equal to the comptime value, I don't see how we are going to have an issue. Maybe I'm missing something?

Reply:
Oops, you're right. I just read it wrongly for some reason. I probably thought it was checking a receiver (again). I'm sorry!